### DISCUSSION TITLE
Vite plugin

### DISCUSSION THREAD

**Discussion by r1m** [terrazzoapp/terrazzo#167] [Q&A]
Hello,
I would like to set up this inside my tool chain using vite. However looking at the code it feel like I would have to rewrite most of the cli logic to parse config and apply plugin.
Is there a better way to use cobalt from node code?

**Comment by drwpow** [comment]
Thanks for the suggestion! I’m a big fan of Vite, and I’d be open to writing a plugin, but I’m not sure if the workflow would be viable for many setups.

Vite plugins require a 1:1 relationship. A Vite plugin would let you import your `tokens.json` as JS (basically pretend as if it were generated by `@cobalt-ui/plugin-js`). But that would limit you to JS and only JS. If you wanted to use CSS, Sass, and JSON in different parts, you’d still have to use the CLI anyway to generate those.

Alternatively, if the Vite plugin tried to assume output for all current Cobalt plugins, it would probably not work as expected, since other Vite plugins would fight for control ([see Rollup’s `resolve()` option](https://rollupjs.org/plugin-development/#resolveid) where plugins “claim” ownership for specific files/formats). There would be unexpected behavior trying to run some CSS through Cobalt and other CSS through Vite’s normal pipelines that would probably just be simpler keeping the two separate like they are currently.

There’s also other plugins like [Tailwind](https://cobalt-ui.pages.dev/integrations/tailwind) that couldn’t operate as part of the normal Vite pipeline, either.

But maybe you envision something different? If you’ve developed Rollup/Vite plugins before and are describing something different, would love feedback / open to other suggestions.

**Comment by Siilwyn** [comment]
For now I'm using something like this:
```mts
import { defineConfig, type Plugin } from 'vite';

import process from 'node:process';
import { promisify } from 'node:util';
import { exec } from 'node:child_process';
import { fileURLToPath } from 'node:url';

const execPromise = promisify(exec);

export default defineConfig({
	plugins: [
		tokensBuildPlugin(),
	],
});

function tokensBuildPlugin(): Plugin {
	function runTerrazzoBuild() {
		console.info('Running terrazzo build...');
		return execPromise('tz build')
			.then(({ stdout, stderr }) => {
				stdout && console.info(stdout);
				stderr && console.error(stderr);
			})
			.catch((error) => console.error(error.message));
	}

	return {
		name: 'tokens-build-watcher',
		configureServer(server) {
			const tokensPath = fileURLToPath(
				new URL('./src/tokens.json', import.meta.url),
			);

			server.watcher.add(String(tokensPath));

			server.watcher.on('change', (changedPath) => {
				if (changedPath === tokensPath) {
					return runTerrazzoBuild();
				}
			});
		},
		buildStart() {
			return runTerrazzoBuild();
		},
	};
}
```
